### PR TITLE
[WIP] New frontend with angular 14+

### PR DESCRIPTION
**Pull request by inpercima**
This is a very first preview of the lvzviz, how it could look with a new or updated frontend framework, angular 13. So this is a working draft.

I found some issues and ideas to change the layout for improvements.
I took the liberty of first creating a separate area under /client to work independent on this feature.

The current status can be viewed at http://lvzviz.inpercima.net. It is just the frontend without reference to the backend. The data used are stored statically here.

In the branch in the client folder, a mock server (npm run run:mock) and the angular app (npm run serve:mock) can currently be started.

The branch contains:
* angular 13.0.2
* leaflet 1.7.1
* angular material 13.0.2
* angular-split 13.0.1

What did you think about this update? Would we do this?
Please give me feedback what you think of it.

### CODE REVIEW

**Comment by inpercima** [comment]
@sepe81: You did the most changes in the last year. What did you think about this change? Would you like to realize this feature change?

**Comment by sepe81** [comment]
Hey @inpercima -- thank you for the comprehensive work. Migrating to Angular would be a great step. I'll try to review it in the following weeks.

File: client/cypress/integration/spec.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+describe('My First Test', () => {
+  it('Visits the initial project page', () => {
+    cy.visit('/')
+    cy.contains('Hello world')
```
Comment (sepe81) [inline client/cypress/integration/spec.ts:4]: Could this become a real test (no hello world)?

File: client/src/app/app.component.html [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<div>
+  <mat-toolbar color="primary" class="mat-elevation-z4">
+    {{ appname }}
+  </mat-toolbar>
+  <mat-tab-group mat-align-tabs="start">
+    <mat-tab label="Suche">
+      <lvzviz-search></lvzviz-search>
+    </mat-tab>
+    <mat-tab label="Statistik">
```
Comment (sepe81) [inline client/src/app/app.component.html:9]: Is this the tab for the heatmap with timeline feature?

File: client/.editorconfig [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Editor configuration, see https://editorconfig.org
```
Comment (sepe81) [inline client/.editorconfig:1]: Could this be part of the real root config?

Comment (sepe81) [reply client/.editorconfig:1]: same for .gitattributes, .gitignore, ...

**Review by sepe81 — CHANGES_REQUESTED** [review]

File: client/src/app/app.component.html (hunk 2 above)
Comment (inpercima) [reply client/src/app/app.component.html:9]: Yes, this is the preparation for the heatmap if it continues to be used.

File: client/.editorconfig (hunk 3 above)
Comment (inpercima) [reply client/.editorconfig:1]: Of course, I just generated all to see what will be used and as first part separated from the main source.

File: client/cypress/integration/spec.ts (hunk 1 above)
Comment (inpercima) [reply client/cypress/integration/spec.ts:4]: Yes, like described I just created a minimal prototype to check if this feature is welcome. ;)

**Comment by inpercima** [comment]
Next steps for me:
* work with docker image of elasticsearch to get full response and check pagination
* integrate heatmap

File: client/src/app/features/statistic/statistic.component.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    var latlng: any[] = [];
+    content.forEach((c) => {
+      if (c.coords) {
+        latlng.push([c.coords.lat, c.coords.lon]);
+      }
+    });
+    this.heat.setLatLngs(latlng);
+    // TODO: what does this line do?
+    // currently now functional b/c _size is not defined
+    //if (this.map._size.y === 0) { } else {
```
Comment (inpercima) [inline client/src/app/features/statistic/statistic.component.ts:102]: I am not sure about this line. Currently I commented it out b/c this gives "_size is undefined".
What does this line do?
Has this anything to do with the coords of the points?
There are currently differences.
heatmap

Comment (inpercima) [reply client/src/app/features/statistic/statistic.component.ts:102]: I changed the code to
```
if (this.map.getSize().y !== 0) {
      this.map.addLayer(this.heat);
    }
```
But I don't know what this check does.

For the heatmap:
I found a reason why the maps are different. The results are different, on localhost the most articles are snippets only. The server seems to load and get the data in a loop so most articles are fulltext and so more words matches.

**Comment by inpercima** [comment]
> Next steps for me:
> 
>     * work with docker image of elasticsearch to get full response and check pagination
> 
>     * integrate heatmap

These steps are closed.

Next steps for me:
* adjust unit and e2e tests
* look to update git-files (.gitattributes, .gitignore) and .editorconfig
* adjustments for mobile view
* look how to integrate both, backend and frontend in this project

**Comment by inpercima** [comment]
Now it comes to the last point: Merging the new front end with the back end.

A way would be to really separate the two and create a folder for the frontend and one for the backend and then combine them with each other.

I would suggest leaving the frontend in the web folder and tidying up what is no longer needed in the root and making all the adjustments so that the frontend is integrated into the app as before.

**Comment by inpercima** [comment]
Now I'm at a point where I need some help.

Two things: Elasticsearch is currently starting up with a problem, but it's running. I cannot discuss the problem in detail. Here's the bug.
```
2022-01-26 21:14:34.257 ERROR 23018 --- [   scheduling-1] o.s.s.s.TaskUtils$LoggingErrorHandler    : Unexpected error occurred in scheduled task

org.springframework.data.elasticsearch.ElasticsearchException: Bulk indexing has failures. Use ElasticsearchException.getFailedDocuments() for detailed messages [{41a77a6091cc863cb7986ef8514b58f603904e09=ElasticsearchException[Elasticsearch exception [type=cluster_block_exception, reason=blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];]]
```

The part with `FORBIDDEN/12/index read-only` is repeating more times with an other message id.

The second one is the build itself.
There are many Dockerfiles and gradle options I do not know in detail for this project.
The angular-app can be build with `ng build:prod` and results in `./web/dist` and is complete but I am not sure how I can integrate it correct.

**Comment by inpercima** [comment]
Now angular 14 is included.

**Comment by inpercima** [comment]
@sepe81 Now the new frontend using the new backend (version 2.1.0) is temporarily online under http://lvzviz.inpercima.net/

**Comment by inpercima** [comment]
@sepe81 Tell me if you have some time. I will make sure that the system is then running to check.

**Comment by inpercima** [comment]
Angular is still updating and so many months ago I didn't update this PR so i decided to create a new one with learnings by this feature and PR. Find the new one under: https://github.com/CodeforLeipzig/lvz-viz/pull/153

**Pull request closed without merging**